### PR TITLE
Enumerable methods TSE Review.

### PR DESCRIPTION
**Pull request by carlosveigadev**
The following methods were created and tested.
```
#my_each;
#my_each_with_index;
#my_select;
#my_all?;
#my_any;
#my_none;
#my_count;
#my_map;
#my_inject;
```

I did this project alone because my partner stopped answering due to connectivity issues, so I was requested by Willow to do it by myself.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    puts "#{__FILE__}:#{__LINE__}: w: block not used" if arg && block_given?
+    return my_all_extra_func(arg) if arg
+
+    if !block_given?
+      size.times { |n| return false if to_a[n] == false || to_a[n].nil? }
+    else
+      my_each { |n| return false unless yield n }
+    end
+    true
+  end
```
Comment (christianotieno) [inline main.rb:36]: # my_all? 
1. [ ] When a Regex is passed as an argument returns false if any of the collection does not match the Regex
### Test 👉  `words.my_all?(/o/)`

|Expected|Got|
|:---:|:---:|
|false|true|

#
2. [ ] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
### Test 👉  `array.my_all?(3)`

|Expected|Got|
|:---:|:---:|
|true|`TypeError`|
#

3. [ ] When a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern.
### Test 👉  `array.my_all?(3)`

|Expected|Got|
|:---:|:---:|
|true|`TypeError`|

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    if arg
+      puts "#{__FILE__}:#{__LINE__}: w: block not used" if block_given?
+      return my_any_extra_func(arg)
+    elsif block_given?
+      my_each { |n| return true if yield n }
+    else
+      my_each { |n| return true if n }
+    end
+    false
+  end
```
Comment (christianotieno) [inline main.rb:58]: # my_any?
1. [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class:: Numeric

### Test 👉  `array.my_any?(Numeric)`

|Expected|Got|
|:---:|:---:|
|true|`#<Array:2080> => [2, 4, 7, 3, 7, 1, 3, 5, 7, 7, 4, 4, 2, 7, 3, 3, 4, 2, 5, 2, 1, 3, 4, 5, 8, 2, 6, 0, 5, 0, 0, 4, 5, 1..., 5, 8, 4, 5, 2, 4, 8, 2, 3, 3, 4, 5, 6, 0, 4, 5, 7, 3, 3, 2, 6, 0, 3, 1, 8, 1, 1, 2, 1, 8, 8, 6, 0]`|

#

2. [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex

### Test 👉 ` words.my_any?(/d/)`

|Expected|Got|
|:---:|:---:|
|true|`#<Array:2100> => ["dog", "door", "rod", "blade"]`|

#
3. [ ]  When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
### Test 👉  `words.my_any?('cat')`

|Expected|Got|
|:---:|:---:|
|true|false|

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  def my_none?(arg = nil)
+    if arg
+      puts "#{__FILE__}:#{__LINE__}: w: block not used" if block_given?
+      !my_any?(arg)
+    elsif block_given?
+      !my_any?(&Proc.new)
+    else
+      !my_any?
+    end
+  end
```
Comment (christianotieno) [inline main.rb]: # my_none
1. [ ] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

### Test 👉 `words.my_none?(5)`

|Expected|Got|
|:---:|:---:|
|false|true|

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    counter = 0
+    if block_given?
+      my_each { |n| counter += 1 if yield n }
+    elsif arg.zero?
+      my_each { counter += 1 }
+    else
+      my_each { |n| counter += 1 if arg == n }
+    end
+    counter
+  end
```
Comment (christianotieno) [inline main.rb:90]: # my_count
1. [ ] Counts the number of items in enum that are equal to the item if an argument is given
### Test 👉 `array.my_count(LOWEST_VALUE)`

|Expected|Got|
|:---:|:---:|
|`9`|`100`|

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    if sym
+      arr.my_each { |n| accum = accum ? accum.send(sym, n) : n }
+    elsif block_given?
+      arr.my_each { |n| accum = accum ? yield(accum, n) : n }
+    else
+      raise 'no block given'
+    end
+    accum
+  end
+end
```
Comment (christianotieno) [inline main.rb]: # my_inject
1. [ ] Raises a `"LocalJumpError"` when no block or argument is given 
### Test 👉 ` array.my_inject`

|Expected|Got|
|:---:|:---:|
|`LocalJumpError`|`#<RuntimeError: no block given> with backtrace:`|

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @wrakc   :wave:
-
You have done a splendid job! :tada: Keep it up.
We, however, do have some few changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Comment by carlosveigadev** [comment]
Thanks for the support @christianotieno!

I improved the code as requested:
- Fixed my_all? not correctly returning a boolean value when a regex was passed as an argument;
- Fixed my_all? TypeError when an arg was passed;
- Fixed my_any? returning self when class was passed as an argument or regex;
- Fixed my_any? not returning a right boolean value when a pattern was passed as an argument;
- Fixed my_none? not returning a right boolean value when a pattern was passed as an argument;
- Fixed my_count to count the similar values when a number was passed as an argument;
- Fixed my_inject to return LocalJumpError when a block was not given to the method.

**Comment by addod19** [comment]
## Pull Request under review by Daniel Larbi Addo (TSE)

**Comment by addod19** [comment]
## Pull Request under review by Daniel Larbi Addo (TSE)

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @wrakc, great job with this project :+1: However, there are a few changes we need to implement in order to approve the project. Kindly check comments below: :smile: 
- ```test.rb``` kindly change the ```require``` to ```require_relative```. Since ```require_relative``` is used for files in our current directory
- I test the methods, and this is the result as we compare to the built-in enumerable methods. The screenshot below will help us address these test cases
- There is warning with ```my_none``` method but returns ```true``` as in the ```none``` built-in enumerable method
NONE
NONE1


Kindly fix these and submit another code review :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo)

**Comment by carlosveigadev** [comment]
Thanks for the support @addod19.

I improved the code as requested:
- Change require to require_relative in the test.rb;
- Fixed the warning about Proc.new deprecated.

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### You got it right this time :clap: the project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next project.
- Happy coding!

**Pull request merged**